File: runaway/src/main/java/dohun/kim/runaway/state/StateGenerator.java
package dohun.kim.runaway.state;

import java.util.List;

import javax.lang.model.element.Element;

import dohun.kim.runaway.exception.AlreadyTakenStateException;

public abstract class StateGenerator {

    private Element element;

    private StateGenerator() {
    }

    public StateGenerator(Element element) {
        this.element = element;
    }

    public abstract State generateState(List<State> existStates) throws AlreadyTakenStateException;

    public Element getElement() {
        return element;
    }
}

File: runaway/src/main/java/kr/co/finda/runaway/state/StateGenerator.kt
package kr.co.finda.runaway.state

import kr.co.finda.runaway.getTypeName
import java.util.*
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement

fun generateState(element: Element, existStates: List<State>): State {
    val state = generateMethodState(element)

    val isAlreadyExistState = existStates.any { it.name == state.name }
    if (isAlreadyExistState) {
        throw IllegalStateException()
    }

    return state
}

private fun generateMethodState(element: Element): State {
    val typeName = (element as ExecutableElement).getTypeName()
    val name = element.simpleName.toString()
        .replace("get", "")
        .replace("set", "")
        .decapitalize(Locale.getDefault())

    return State(typeName, name)
}